scarab/service.go

```diff
@@ -0,0 +1,354 @@
+	Proof     ol.Proof
+	Ephemeral kyber.Point
+	Signature *darc.Signature
+}
+
+// CreateLTS takes as input a roster with a list of all nodes that should
+// participate in the DKG. Every node will store its private key and wait
+// for decryption requests.
+// This method will create a random LTSID that can be used to reference
+// the LTS group created.
+func (s *Service) CreateLTS(cl *CreateLTS) (reply *CreateLTSReply, err error) {
+	tree := cl.Roster.GenerateNaryTreeWithRoot(len(cl.Roster.List), s.ServerIdentity())
+	pi, err := s.CreateProtocol(protocol.NameDKG, tree)
+	setupDKG := pi.(*protocol.SetupDKG)
+	setupDKG.Wait = true
+	reply = &CreateLTSReply{LTSID: make([]byte, 32)}
+	random.New().XORKeyStream(reply.LTSID, reply.LTSID)
+	setupDKG.SetConfig(&onet.GenericConfig{Data: reply.LTSID})
+	log.Lvlf3("%s: reply.LTSID is: %x", s.ServerIdentity(), reply.LTSID)
+	if err := pi.Start(); err != nil {
+		return nil, err
+	}
+	log.Lvl3("Started DKG-protocol - waiting for done", len(cl.Roster.List))
+	select {
+	case <-setupDKG.SetupDone:
+		shared, err := setupDKG.SharedSecret()
+		if err != nil {
+			return nil, err
+		}
+		s.saveMutex.Lock()
+		s.storage.Shared[string(reply.LTSID)] = shared
+		dks, err := setupDKG.DKG.DistKeyShare()
+		if err != nil {
+			s.saveMutex.Unlock()
+			return nil, err
+		}
+		s.storage.Polys[string(reply.LTSID)] = &pubPoly{s.Suite().Point().Base(), dks.Commits}
+		s.storage.Rosters[string(reply.LTSID)] = &cl.Roster
+		s.saveMutex.Unlock()
+		reply.X = shared.X
+	case <-time.After(propagationTimeout):
+		return nil, errors.New("dkg didn't finish in time")
+	}
+	return
+}
+
+// DecryptKey takes as an input a Read- and a Write-proof. Proofs contain
+// everything necessary to verify that a given instance is correct and
+// stored in omniledger.
+// Using the Read and the Write-instance, this method verifies that the
+// requests match and then re-encrypts the secret to the public key given
+// in the Read-instance.
+// TODO: support ephemeral keys.
+func (s *Service) DecryptKey(dkr *DecryptKey) (reply *DecryptKeyReply, err error) {
+	reply = &DecryptKeyReply{}
+	log.Lvl2("Re-encrypt the key to the public key of the reader")
+
+	var read Read
+	if err := dkr.Read.ContractValue(cothority.Suite, ContractReadID, &read); err != nil {
+		return nil, errors.New("didn't get a read instance: " + err.Error())
+	}
+	var write Write
+	if err := dkr.Write.ContractValue(cothority.Suite, ContractWriteID, &write); err != nil {
+		return nil, errors.New("didn't get a write instance: " + err.Error())
+	}
+	if !read.Write.Equal(ol.NewInstanceID(dkr.Write.InclusionProof.Key)) {
+		return nil, errors.New("read doesn't point to passed write")
+	}
+	s.saveMutex.Lock()
+	roster := s.storage.Rosters[string(write.LTSID)]
+	if roster == nil {
+		s.saveMutex.Unlock()
+		return nil, errors.New("don't know the LTSID stored in write")
+	}
+	s.saveMutex.Unlock()
+
+	// Start OCS-protocol to re-encrypt the file's symmetric key under the
```

Comment:
s/OCS/SCARAB/?

Reply:
No - because Scarab is the system comprised of AC + SC. OCS here is the protocol doing the re-encryption. So it should be `s/OCS/ReEncrypt/` or something.